### PR TITLE
Better payload stats

### PR DESCRIPTION
**Pull request by kenoir**
For: https://github.com/wellcomecollection/platform/issues/4521

### CODE REVIEW

File: bag_replicator/src/main/scala/uk/ac/wellcome/platform/archive/bagreplicator/services/BagReplicatorWorker.scala [hunk 1]
```diff
@@ -72,10 +69,11 @@ class BagReplicatorWorker[
...
 
-      dstPrefix = destinationBuilder.buildDestination(
-        storageSpace = payload.storageSpace,
-        externalIdentifier = payload.externalIdentifier,
-        version = payload.version
+      dstPrefix = DestinationBuilder.buildDestination(
+        namespace = destinationConfig.namespace,
+        payload.storageSpace,
+        payload.externalIdentifier,
+        payload.version
```
Comment (alexwlchan) [inline bag_replicator/src/main/scala/uk/ac/wellcome/platform/archive/bagreplicator/services/BagReplicatorWorker.scala:76]: nit: put the keyword arguments back please?

File: common/src/test/scala/uk/ac/wellcome/platform/archive/common/fixtures/StorageRandomThings.scala [hunk 2]
```diff
@@ -185,6 +185,9 @@ trait StorageRandomThings extends RandomThings {
 
   def createBagPath: BagPath = BagPath(randomAlphanumeric)
 
+  def createBagPathWithPrefix(prefix: String, name: String): BagPath =
+    BagPath(f"$prefix/$name")
```
Comment (alexwlchan) [inline common/src/test/scala/uk/ac/wellcome/platform/archive/common/fixtures/StorageRandomThings.scala:189]: nit: we tend to use s-formatted strings elsewhere. Is there a reason you've started to switch to f-strings?

File: common/src/test/scala/uk/ac/wellcome/platform/archive/common/storage/services/DestinationBuilderTest.scala [hunk 3]
```diff
@@ -21,48 +21,41 @@ class DestinationBuilderTest
...
-
-    val location = builder.buildDestination(
-      storageSpace = StorageSpace("alfa"),
-      externalIdentifier = ExternalIdentifier("bravo/charlie"),
-      version = version
+    val location = DestinationBuilder.buildDestination(
+      namespace = namespace,
+      StorageSpace("alfa"),
+      ExternalIdentifier("bravo/charlie"),
+      version
```
Comment (alexwlchan) [inline common/src/test/scala/uk/ac/wellcome/platform/archive/common/storage/services/DestinationBuilderTest.scala:58]: nit: keyword arguments?

File: indexer/bag_indexer/src/main/scala/uk/ac/wellcome/platform/archive/indexer/bags/services/FileSuffixCounter.scala [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+object FileSuffixCounter {
+  def count(files: Seq[StorageManifestFile]): Map[String, Int] = {
+    files
+      .map(_.name.split("\\."))
+      .filter(_.length > 1)
+      .map(_.reverse.head)
+      .groupBy(identity)
+      .mapValues(_.size)
+  }
```
Comment (alexwlchan) [inline indexer/bag_indexer/src/main/scala/uk/ac/wellcome/platform/archive/indexer/bags/services/FileSuffixCounter.scala:13]: 😻 

FP style makes this very nice.

Two suggestions:

* Can you replace `.map(_.reverse.head)` with `.map(_.last)`?
* Should we lowercase the suffix? e.g. if a bag has `cat.jp2` and `dog.JP2`, we should combine those together into a single entry.

File: indexer/bag_indexer/src/main/scala/uk/ac/wellcome/platform/archive/indexer/bags/models/IndexedStorageManifest.scala [hunk 5]
```diff
@@ -38,31 +50,34 @@ case class IndexedStorageManifest(
...
   def apply(storageManifest: StorageManifest): IndexedStorageManifest = {
-    val payloadFiles = storageManifest.manifest.files.map(IndexedFileFields(_))
+    val storageManifestFiles = storageManifest.manifest.files
+    val payloadFiles = storageManifestFiles.map(IndexedFileFields(_))
+
+    val fileSuffixTally = FileSuffixCounter.count(storageManifestFiles)
+    val indexedSuffixTally = fileSuffixTally
+      .map(IndexedSuffixTally(_))
+      .toList
+      .sortBy(_.suffix)
```
Comment (alexwlchan) [inline indexer/bag_indexer/src/main/scala/uk/ac/wellcome/platform/archive/indexer/bags/models/IndexedStorageManifest.scala:66]: I'd sort by count instead, so when I'm looking in the reporting cluster the most common file types float to the top.

File: indexer/bag_indexer/src/test/scala/uk/ac/wellcome/platform/archive/indexer/bag/models/IndexedStorageManifestTest.scala [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+  IndexedPayloadStats,
+  IndexedStorageManifest,
+  IndexedSuffixTally
+}
+
+class IndexedStorageManifestTest
+    extends AnyFunSpec
+    with Matchers
+    with IngestGenerators
+    with StorageManifestGenerators {
```
Comment (alexwlchan) [inline indexer/bag_indexer/src/test/scala/uk/ac/wellcome/platform/archive/indexer/bag/models/IndexedStorageManifestTest.scala:22]: How about a short test that it pulls through other attributes like "id" and "space"?

File: indexer/bag_indexer/src/test/scala/uk/ac/wellcome/platform/archive/indexer/bag/services/FileSuffixCounterTest.scala [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    createStorageManifestFileWith(
+      name = "bad..bmp"
+    ),
+    createStorageManifestFileWith(
+      name = "worse.jif."
+    ),
+    createStorageManifestFileWith(
+      name = "file.with.full.stops.scala"
+    )
+  )
```
Comment (alexwlchan) [inline indexer/bag_indexer/src/test/scala/uk/ac/wellcome/platform/archive/indexer/bag/services/FileSuffixCounterTest.scala:38]:
```suggestion
  val names = Seq("foo.txt", "bar.txt", "bat.png", "mystery file", "bad..bmp", "worse.jif.", "file.with.full.stops.scala")
  val files = names.map { createStorageManifestFileWith(_) }
```

File: indexer/bag_indexer/src/test/scala/uk/ac/wellcome/platform/archive/indexer/bag/services/FileSuffixCounterTest.scala [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+  it("works") {
```
Comment (alexwlchan) [inline indexer/bag_indexer/src/test/scala/uk/ac/wellcome/platform/archive/indexer/bag/services/FileSuffixCounterTest.scala:40]:
```suggestion
  it("tallies the file suffixes correctly") {
```

File: bag_replicator/src/main/scala/uk/ac/wellcome/platform/archive/bagreplicator/services/BagReplicatorWorker.scala (hunk 1 above)
Comment (kenoir) [reply bag_replicator/src/main/scala/uk/ac/wellcome/platform/archive/bagreplicator/services/BagReplicatorWorker.scala:76]: Got lost as this was temporarily curried, where keyword args are not allowed.

File: common/src/test/scala/uk/ac/wellcome/platform/archive/common/fixtures/StorageRandomThings.scala (hunk 2 above)
Comment (kenoir) [reply common/src/test/scala/uk/ac/wellcome/platform/archive/common/fixtures/StorageRandomThings.scala:189]: 🐍 !

File: indexer/bag_indexer/src/main/scala/uk/ac/wellcome/platform/archive/indexer/bags/services/FileSuffixCounter.scala (hunk 4 above)
Comment (kenoir) [reply indexer/bag_indexer/src/main/scala/uk/ac/wellcome/platform/archive/indexer/bags/services/FileSuffixCounter.scala:13]: Me while writing this "I wish I could just get the last element of a list".

Comment (kenoir) [reply indexer/bag_indexer/src/main/scala/uk/ac/wellcome/platform/archive/indexer/bags/services/FileSuffixCounter.scala:13]: Lowercasing is a definite yes.

File: indexer/bag_indexer/src/main/scala/uk/ac/wellcome/platform/archive/indexer/bags/models/IndexedStorageManifest.scala (hunk 5 above)
Comment (kenoir) [reply indexer/bag_indexer/src/main/scala/uk/ac/wellcome/platform/archive/indexer/bags/models/IndexedStorageManifest.scala:66]: reverse count then I think - so highest first.

**Pull request merged**